### DISCUSSION TITLE
Unable to resolve error: Bad_TcpMessageTooLarge max chunk count exceeded (64)

### DISCUSSION THREAD

**Discussion by bsalmilo** [eclipse-milo/milo#1277] [General]
Hi guy, apologies if this has been discussed already. I've been combing through documentation and various forums but I haven't been able to figure out what the issue is. The full error: 

```
ERROR org.eclipse.milo.opcua.stack.client.transport.uasc.UascClientMessageHandler - Exception caught: UaException: status=Bad_TcpMessageTooLarge, message=max chunk count exceeded (64). 
```

This occurs when I try to call the browse function on a node, as described here - https://javadoc.io/static/org.eclipse.milo/sdk-client/0.3.3/org/eclipse/milo/opcua/sdk/client/api/services/ViewServices.html#browse-org.eclipse.milo.opcua.stack.core.types.structured.BrowseDescription-. 

My research led to me this forum - https://www.eclipse.org/lists/milo-dev/msg00967.html which told me to try using MaxReferencesPerNode toggle. This meant I had to change my request to the other browse method, which returns a BrowseResponse as opposed to a BrowseResult. Which is fine. However this led me to a different issue. 

```
stdout. UaException: status=Bad_Timeout, message=requestId=4 timed out after 60000ms.
```

Seems to not be able to perform the browsing in time anymore. 

Not sure where to go from here as information on this seems pretty limited so any help would be fantastic. Thanks.

**Comment by kevinherron** [comment]
It would be useful if you could get a Wireshark capture and upload it here. It seems the server you’re connected to isn’t responding in time but I’d like to verify.

**Reply by kevinherron** [reply]
Some code snippets that demonstrate what you’re doing in your client code would be useful as well.

**Reply by bsalmilo** [reply]
The way I was doing it originally was building a browse description and browsing for it. 

```
val browseDescription = BrowseDescription(
                miloNodeId,
                BrowseDirection.Both,
                referenceTypeToRetrieve,
                INCLUDE_SUBTYPES,
                ALL_NODE_CLASSES_RESULT_MASK,
                Unsigned.uint(BrowseResultMask.All.value)
            )

            coroutineScope {
                    async {
                        client.browse(browseDescription)
                            .asDeferred()
                            .await()
                            ?.let { it.references?.filterNotNull() }
                            ?: emptyList()
                    }
                }
```
This returned the original error (max chunk count exceeded) but there was nothing of interest in the Wireshark capture, indicating it was an issue on the client side. I then changed the browsing logic to the following: 

```
coroutineScope {
                    async {
                        client.browse(null, UInteger.valueOf(4000), listOf(browseDescription))
                            .asDeferred()
                            .await()
                            ?.let { it.results.flatMap { t -> t.references.filterNotNull() } }
                            ?: emptyList()
                    }
                }
```
I was trying to add the maxReferences limit here, as suggested by the forum post I linked above, but it times out now (over 60000ms). 

Same thing here - nothing of any value shows up in the Wireshark capture.

**Reply by kevinherron** [reply]
>This returned the original error (max chunk count exceeded) but there was nothing of interest in the Wireshark capture, indicating it was an issue on the client side.

That's not quite the right conclusion. The error message means the server is trying to send a response larger than the negotiated maximum size. You either need to increase the message size, or do something different in your request to reduce the number of references returned, which seems to be the route you've chosen. 

You may need to do some combination of reducing the number of nodes your browse at once (number of BrowseDescriptions in the call) and/or reducing the max references per node.

**Reply by bsalmilo** [reply]
> You either need to increase the message size, or do something different in your request to reduce the number of references returned, which seems to be the route you've chosen.You may need to do some combination of reducing the number of nodes your browse at once (number of BrowseDescriptions in the call) and/or reducing the max references per node.

I believe that there is only one browse description in the call. Originally, I was calling the version of browse that "browses a singular node." After the update, I believe it still only browsed that singular node. 

What would you advise setting the maxReferencesPerNode to?

**Reply by kevinherron** [reply]
You could just try 1000 and see if it's your issue or not. I don't like guessing what's going on here, I'd rather look at a capture.

**Reply by bsalmilo** [reply]
Seems to still get the same timeout issue. I spoke to my security team and unfortunately I don't think I'm able to provide a capture - but I don't see anything notable in there anyway. Is there anything else I can maybe provide?

**Reply by kevinherron** [reply]
No, not really. That timeout error means the server isn't responding within the default 60s timeout period the client waits for a response. 

You'll just have to look closely at a capture that results in a response vs a capture that results in a timeout and try to figure out what's different.

**Reply by bsalmilo** [reply]
From what I can see when filtering for OpcUa in a correct response there are many BrowseRequest/BrowseResponses that populate, but in the failure there is only 1. Other than that, there's no indication of a timeout taking place.

**Comment by bsalmilo** [comment]
Update - I manually increased the max chunk count to 128 in the client creator and that seemed to resolve the issue. My question is - is this an appropriate solution? Are there any ramifications of doing this down the line? @kevinherron

**Reply by kevinherron** [reply]
It's fine, you're just increasing the max message size. It might be a bandaid if you end up later eliciting a larger response from the server that still doesn't fit, but I'm completely guessing here because you weren't allowed to provide me anything useful I could use to see what's actually happening.

**Reply by bsalmilo** [reply]
So this max chunk count error would only happen when we were browsing a large number of nodes - over 10,000. Is there a reason this might happening? Is it just that as you browse more nodes the message returned from the server gets larger? If that's the case - can it be assumed that doubling the encoding limit to 128 will mean that the number of nodes we can safely browse will roughly double? What's the implication of setting the limit to MAX_INTEGER?

**Reply by kevinherron** [reply]
> Is there a reason this might happening? Is it just that as you browse more nodes the message returned from the server gets larger? 

Yes.

You can increase this by playing around with the chunk count, chunk size and max message size values. See https://github.com/eclipse/milo/blob/master/opc-ua-stack/stack-core/src/main/java/org/eclipse/milo/opcua/stack/core/channel/EncodingLimits.java for how they are roughly related.

At a certain point accepting many chunks or a message too large becomes pointless because the server may not have buffers large enough to send the response.

There's no hard and fast rule, you'll have to experiment, or segment/chunk your browses so they stay under a certain size if possible. You only need to increase the encoding limits if those 10,000 nodes are all children of one single node _and_ you haven't limited the max references returned. 

If that's not the case you're doing this to yourself by sending a browse request for many nodes, each of which can have up to max references children, and it's just too much data to fit into a response at the default encoding limits.

**Reply by bsalmilo** [reply]
Might help for me to have a better understanding of what exactly a message is. Is message = single Milo API response? or message = all data received within a single OPC UA connection?

**Reply by kevinherron** [reply]
Each Request and Response is a message.

**Reply by bsalmilo** [reply]
Is there a standard message size for OPC UA servers or does it differ from machine to machine?

**Reply by kevinherron** [reply]
No, there is no standard size. Every client or server may have a different configuration.

**Reply by bsalmilo** [reply]
Is there a way to query the server for this size to set the client's encoding limits to correspond?

**Reply by kevinherron** [reply]
No, it doesn't work that way. The client tells the server its limits in the opening Hello message. The server tells the client its limits in the Acknowledge message. These values aren't reported to the application layer in any way.

Servers can report the "max operations per" for each service in the address space, but even then there are corner cases where you send a request within those limits and it generates a response that is too large for one of the parties.

You need to just code your client somewhat defensively when it comes to the service calls it makes. Keep them reasonably sized, especially the ones that can blow up in size like Browse, and have some way to configure the size of the calls at runtime.

**Reply by bsalmilo** [reply]
Sorry for the ongoing discussion on this. Really appreciate all your help. I tried implementing some kind of pagination by using BrowseNext instead of browse, but still running into the same chunk issue. Do you happen to have any other suggestions beyond increasing the encoding limits?

**Reply by kevinherron** [reply]
How many BrowseDescriptions are you passing to each Browse service call? Limiting this should limit the size of the BrowseResponse. If you're only passing 1, then limiting the maxReferencesPerNode should limit the size. If this doesn't work, the server may be ignoring the parameter and there's nothing you can do except increase encoding limits. A Wireshark capture would help a lot here.

**Reply by bsalmilo** [reply]
The node that fails has only 1 reference variable node. Is this because of the historic accumulation of the previous nodes that the message becomes too large?

**Reply by kevinherron** [reply]
Previous service calls should not affect new ones, so unless you mean previous nodes within the same BrowseRequest, not sure what is happening.
